### ISSUE TITLE
Review block verification logic

### ISSUE THREAD

**Issue by karmacoma** [LiskHQ/lisk#459]
Belongs to: #449 

Previous task list:

- [x] Provide single entry point for block verification
- [x] Improve block verifications/checks, low-cost ones should be done first, heavy ones - last
- [x] Break block verification on first error (performance, we don't need to complete all other checks if one failed and block is already invalid)
- [x] Eliminate any double checks

New task list: 

- [ ] Provide single entry point for block verification
- [ ] Verify blocks for fork during verify, instead of onReceiveBlock
- [ ] Reorder the operations in `blocks/verify.js`, processBlock - https://github.com/LiskHQ/lisk/blob/development/modules/blocks/verify.js#L325

1. Normalize
2. Verify
3. Broadcast
4. Apply block

`checkExists`, `validateBlockSlot` and `checkTransactions` should all be performed inside of the verify logic.

- [ ] Refactor `onReceiveBlock` behaviors to occur during `verify`, instead of occurring at a separate point in time.

1. postBlock: https://github.com/LiskHQ/lisk/blob/development/modules/transport.js#L493
2. onReceiveBlock: https://github.com/LiskHQ/lisk/blob/development/modules/blocks/process.js#L382
These checks should move to verify, and occur as the first checks of verify, onReceiveBlock should serve as the master thread for block receipt, with logic being performed inside of it.

- [ ] Move blockId and Height generation into a new place to allow them to be added by the forger, rather than the recipient

1. Remove this entry from here: https://github.com/LiskHQ/lisk/blob/development/modules/blocks/verify.js#L172
2. Remove this function here: https://github.com/LiskHQ/lisk/blob/development/modules/blocks/verify.js#L287
3. Create a new function that adds the functions when generating the block and attach them then.  These items should not be removed from a block

**Comment by Isabello** [comment]
This issue has been reopened:

Block ID and Height should be included in every block, and not added during other steps (such as block verify).
There is still not a single entry point for blocks verification, verification is occurring in many places along the way.

- [ ] Reorder the operations in `blocks/verify.js`, processBlock - https://github.com/LiskHQ/lisk/blob/development/modules/blocks/verify.js#L325

1. Normalize
2. Verify
3. Broadcast
4. Apply block

`checkExists`, `validateBlockSlot` and `checkTransactions` should all be performed inside of the verify logic.

- [ ] Refactor `onReceiveBlock` behaviors to occur during `verify`, instead of occurring at a separate point in time.

1. postBlock: https://github.com/LiskHQ/lisk/blob/development/modules/transport.js#L493
2. onReceiveBlock: https://github.com/LiskHQ/lisk/blob/development/modules/blocks/process.js#L382
These checks should move to verify, and occur as the first checks of verify, onReceiveBlock should serve as the master thread for block receipt, with logic being performed inside of it.

- [ ] Move blockId and Height generation into a new place to allow them to be added by the forger, rather than the recipient

1. Remove this entry from here: https://github.com/LiskHQ/lisk/blob/development/modules/blocks/verify.js#L172
2. Remove this function here: https://github.com/LiskHQ/lisk/blob/development/modules/blocks/verify.js#L287
3. Create a new function that adds the functions when generating the block and attach them then.  These items should not be removed from a block

**Comment by karmacoma** [comment]
Task list resolved indirectly through recent `0.9.*` patch releases.
